### ISSUE TITLE
[Java] ClientReconnectingException thrown during reconnect dropping the subscriptionMessages

### ISSUE THREAD

**Issue by kalinkostashki** [eclipse-ditto/ditto-clients#231]
Currently the following scenario may unfold(very rarely) which cannot be handled on the customer side:

1. Gateway service shutdown
2. Ditto client detects the disconnect and wait for reconnection
3. Gateway service gets backup which:
    3.1. Calls handleReconnectionIfEnabled in default(callback exectuor thread) and reconnecting is set to true
    3.2. [doReconnect()](https://github.com/eclipse-ditto/ditto-clients/blob/7af0486d51f4268be065a8dd7ec85a3956ebecca/java/src/main/java/org/eclipse/ditto/client/messaging/internal/WebSocketMessagingProvider.java#L489) is triggered which in turn calls reconnectWithPotentialRetries in another executor thread(connect executor thread)
    3.3. When that happens websocket is recreated and replaced and at this point we synchronize on the websocket client
    3.4. After that reconnecting is set to false
4. What may happen in this case is that Websocket may reconnect before we set reconnecting to false
5. Leaving you with a working client that actually threw a ClientReconnectingException in some thread but that is it.

Not sure how to handle this currently. One idea is to re-use the connectExecutor and check when moving to onConnected if everything is correct. Would roughly look like this in onConnected():
```
CompletableFuture<Boolean>` isConnectionStable = new CompletableFuture<>();
            Runnable checkTask = new Runnable() {
                @Override
                public void run() {
                    if (reconnecting.get()) {
                        isConnectionStable.complete(true);  // Complete the future if flag is true
                    }
                }
            };
            ScheduledFuture<?> future = connectExecutor.scheduleAtFixedRate(checkTask, 0, 50, TimeUnit.MILLISECONDS);
            try {
                if (isConnectionStable.get(100, TimeUnit.MILLISECONDS)) {
                    future.cancel(true);
                    System.out.println("Connection is stable");
                }
            } catch (InterruptedException | ExecutionException e) {
                e.printStackTrace();
            }
```
So basically a bit lost and needing inspiration :)

**Comment by kalinkostashki** [comment]
Here is a draft PR for the issue:
https://github.com/eclipse-ditto/ditto-clients/pull/232
Currently just catching the interrupted exception and haven't decided how to handle.
But please take a look and give your opinion @thjaeckle
